### ISSUE TITLE
Unsupported API stats to help contributors

### ISSUE THREAD

**Issue by re-fox** [fireeye/speakeasy#3]
First off, thanks for releasing this tool.  

I ran speakeasy against the Malpedia corpus (https://malpedia.caad.fkie.fraunhofer.de/) to get a rough estimate of how many samples successfully emulate with/without tossing errors.  While running this test I gathered up a list of the count of unsupported API functions that were causing emulation to halt.  

I understand a github issue may not be the best place to store this information.  It may be useful for someone looking to contribute by going after the highly used API's first.

The result of ~4k samples (truncating results at 20 - the data has a long tail of one-off's)
```
    130 user32.GetKeyboardType
    122 shell32.CommandLineToArgvW
    122 advapi32.ConvertStringSecurityDescriptorToSecurityDescriptorA
    113 msvbvm60.ordinal_100
     91 kernel32.InitializeConditionVariable
     71 advapi32.EventRegister
     57 mfc42.ordinal_1576
     55 kernel32.DeleteFileA
     48 ole32.CoInitialize
     47 kernel32.GetVolumeInformationA
     46 kernel32.SetThreadLocale
     46 kernel32.ReleaseMutex
     46 comctl32.ordinal_17
     42 kernel32.GetSystemInfo
     41 kernel32.HeapValidate
     39 msvcrt.strlen
     36 kernel32.CreateDirectoryW
     33 kernel32.CreateDirectoryA
     32 msvcrt.__wgetmainargs
     32 msvcrt.__p___initenv
     32 kernel32.OutputDebugStringA
     31 msvcrt.??2@YAPAXI@Z
     30 user32.LoadIconA
     30 kernel32.SetEvent
     30 kernel32.QueryPerformanceFrequency
     30 kernel32.lstrcmpA
     30 kernel32.GetTimeZoneInformation
     28 kernel32.SetDefaultDllDirectories
     27 ole32.OleInitialize
     27 msvcr90._initterm_e
     27 advapi32.RegSetValueExA
     26 advapi32.InitializeSecurityDescriptor
     25 msvcrt._snwprintf
     25 gdi32.GetDeviceCaps
     24 user32.GetProcessWindowStation
     23 ole32.CoInitializeEx
     22 shell32.SHGetFolderPathA
     21 user32.GetCursorInfo
     21 shell32.SHGetFolderPathW
     21 kernel32.AddVectoredExceptionHandler
     20 user32.OpenInputDesktop
     20 msvcrt.strcat
     20 msvcrt.memmove
     20 kernel32.IsDBCSLeadByte
```

In addition to the above data, Malpedia publicly posts an API frequency graph: https://malpedia.caad.fkie.fraunhofer.de/stats/api_dll_frequencies

**Comment by drewvis** [comment]
Hey, thanks a lot that is super helpful! I'll be sure to keep this open.

**Comment by re-fox** [comment]
Quite a few changes were updated last night, I re-ran the analysis and posted up the new results.  I'll keep editing this issue some semi-regular basis to keep the list current.

**Comment by Flerov** [comment]
Could I get a hint for where I have to add undocumented functions for emulation support. I really need this to be working for the API ntoskrnl.ZwCreateKey

**Comment by jtbennett-fe** [comment]
@Flerov this blog has the hints you are looking for: https://www.mandiant.com/resources/blog/using-speakeasy-emulation-framework-programmatically-to-unpack-malware.

**Comment by Flerov** [comment]
@JohnLaTwC Thanks for the link do you have some further recommendations to get into the functionality of API hooks implemented in '/speakesy/winenv/api/kernelmode/' specifically for '[...]/ntoskrnl.py' since the function is not supported and a simple return 0x0 wont help for further emulation of the driver I'm facing.
I would like to get into how to add support for the API 'ZwCreateKey' for a successfull emulation

From my report it seems that "ZwCreateKey" gets called and after that there is only one more api being called 'RtlInitUnicodeString'
According to the listed errors my error type is: 'NoneType' object has no attribute 'to_bytes' raised by the emulation's attempt to call "ZwOpenKey"
My Terminal output give me the following hint:"""Invalid memory read (UC_ERR_READ_UNMAPPED)
Unsupported API: ntoskrnl.ZwCreateKey"""

I appreciate any further links to documentation and help regarding my problem many thanks in advance

**Comment by jtbennett-fe** [comment]
@Flerov I recommend reading the documentation on the API: https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-zwcreatekey

It is up to you to decide how much of the API you want to support in emulation to meet your needs.

**Comment by Flerov** [comment]
@jtbennett-fe I added support for ZwCreateKey as far as I was able to.
Seems like it works better to change the create Flag in the code for ZwOpenKey.

TLDR: Finally it works now, I appreciate it alot thanks

**Comment by jtbennett-fe** [comment]
@Flerov awesome, glad to hear it!
